### PR TITLE
Migrate `pallet-sudo` to `pallet!`

### PR DESCRIPTION
**Pull request by ascjones**
Part of #7882.

Converts the `Sudo` pallet to the new pallet attribute macro introduced in #6877.

Following the upgrade guidelines here: https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines.

## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Sudo` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Sudo` pallet.

### CODE REVIEW

File: frame/sudo/src/lib.rs [hunk 1]
```diff
@@ -40,44 +40,55 @@
...
-//!     }
+//! 	}
 //! }
 //! # fn main() {}
 //! ```
 //!
 //! ## Genesis Config
 //!
-//! The Sudo module depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
+//! The Sudo pallet depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
```
Comment (thiolliere) [inline frame/sudo/src/lib.rs:88]:
```suggestion
//! The Sudo pallet depends on the [`GenesisConfig`].
```

File: frame/sudo/src/lib.rs [hunk 2]
```diff
@@ -153,14 +173,18 @@ decl_module! {
 		/// - O(1).
 		/// - The weight of this call is defined by the caller.
 		/// # </weight>
-		#[weight = (*_weight, call.get_dispatch_info().class)]
-		fn sudo_unchecked_weight(origin, call: Box<<T as Config>::Call>, _weight: Weight) -> DispatchResultWithPostInfo {
+		#[pallet::weight((*_weight, call.get_dispatch_info().class))]
+		pub(crate) fn sudo_unchecked_weight(
+			origin: OriginFor<T>,
+			call: Box<<T as Config>::Call>,
+			_weight: Weight
```
Comment (thiolliere) [inline frame/sudo/src/lib.rs:180]: I think code format guidelines say so
```suggestion
			_weight: Weight,
```

File: frame/sudo/src/lib.rs [hunk 3]
```diff
@@ -174,14 +198,17 @@ decl_module! {
 		/// - Limited storage reads.
 		/// - One DB change.
 		/// # </weight>
-		#[weight = 0]
-		fn set_key(origin, new: <T::Lookup as StaticLookup>::Source) -> DispatchResultWithPostInfo {
+		#[pallet::weight(0)]
+		pub(crate) fn set_key(
+			origin: OriginFor<T>,
+			new: <T::Lookup as StaticLookup>::Source
```
Comment (thiolliere) [inline frame/sudo/src/lib.rs:204]:
```suggestion
			new: <T::Lookup as StaticLookup>::Source,
```

File: frame/sudo/src/lib.rs [hunk 4]
```diff
@@ -128,17 +145,20 @@ decl_module! {
...
-		#[weight = {
+		#[pallet::weight({
 			let dispatch_info = call.get_dispatch_info();
 			(dispatch_info.weight.saturating_add(10_000), dispatch_info.class)
-		}]
-		fn sudo(origin, call: Box<<T as Config>::Call>) -> DispatchResultWithPostInfo {
+		})]
+		pub(crate) fn sudo(
+			origin: OriginFor<T>,
+			call: Box<<T as Config>::Call>
```
Comment (thiolliere) [inline frame/sudo/src/lib.rs:154]:
```suggestion
			call: Box<<T as Config>::Call>,
```

**Review by thiolliere — APPROVED** [review]
polkadot doesn't needs migration as not used in polkadot/kusama and westend has unchanged pallet prefix.

**Comment by ascjones** [comment]
CI is failing with this warning as error https://gitlab.parity.io/parity/substrate/-/jobs/867537#L473:

```
error: associated function is never used: `call_functions`
  --> frame/sudo/src/mock.rs:48:12
   |
48 |     #[pallet::call]
   |       ^^^^
```

I don't get it because it is clearly used by the `construct_runtime` expansion: `FnEncode ( logger :: Pallet :: < Test > :: call_functions`. Any ideas @thiolliere?

**Comment by thiolliere** [comment]
yes but the metadata function (which uses call_functions) generated by construct_runtime is unused I guess, thus it complains about unused items.

**Comment by ascjones** [comment]
Ah yeah, I wonder how that worked with the legacy `decl_module!` though

**Comment by thiolliere** [comment]
maybe rustc is not doing the same checks for both macros kind.
Anyway you can add a dummy test:
```rust
#[test] 
fn dummy_use() {
Test::metadata();
}
```
Or actually we can add in construct_runtime `#[allow_unused]` for the function metadata.

**Comment by ascjones** [comment]
> Ah yeah, I wonder how that worked with the legacy `decl_module!` though

The original macro already annotated the generated fn with `#[allow(dead_code)]`, I've just done the same for the new one.

**Review by shawntabrizi — APPROVED** [review]

**Pull request merged**